rustls/src/builder.rs

```diff
@@ -62,81 +61,70 @@ use crate::versions;
 /// This may be shortened to:
 ///
 /// ```
-/// # use rustls::client_config_builder_with_safe_defaults;
+/// # use rustls::ClientConfig;
 /// # let root_certs = rustls::RootCertStore::empty();
 /// # let trusted_ct_logs = &[];
-/// client_config_builder_with_safe_defaults()
+/// ClientConfig::builder()
+///     .with_safe_defaults()
 ///     .with_root_certificates(root_certs, trusted_ct_logs)
 ///     .with_no_client_auth();
 /// ```
 ///
-///
 /// The types used here fit together like this:
 ///
+/// 1. Call [`ClientConfig::builder()`] or [`ServerConfig::builder()`] to initialize a builder.
 /// 1. You must make a decision on which cipher suites to use, typically
-///    by calling [`ConfigWantsCipherSuites::with_safe_default_cipher_suites()`].
+///    by calling [`ConfigBuilder<S, WantsCipherSuites>::with_safe_default_cipher_suites()`].
 /// 2. Now you must make a decision
-///    on key exchange groups: typically by calling [`ConfigWantsKxGroups::with_safe_default_kx_groups()`].
+///    on key exchange groups: typically by calling
+///    [`ConfigBuilder<S, WantsKxGroups>::with_safe_default_kx_groups()`].
 /// 3. Now you must make
 ///    a decision on which protocol versions to support, typically by calling
-///    [`ConfigWantsVersions::with_safe_default_protocol_versions()`].
-/// 4. You now need to indicate whether to make a [`ServerConfig`] or [`ClientConfig`],
-///    by calling [`ConfigWantsPeerType::for_server()`]
-///    or [`ConfigWantsPeerType::for_client()`] respectively.
-/// 5. Now see [`ConfigWantsServerVerifier`] or [`ConfigWantsClientVerifier`] for further steps.
+///    [`ConfigBuilder<S, WantsVersions>::with_safe_default_protocol_versions()`].
+/// 5. Now see [`ConfigBuilder<ClientConfig, WantsVerifier>`] or
+///    [`ConfigBuilder<ServerConfig, WantsVerifier>`] for further steps.
 ///
 /// [`ServerConfig`]: crate::ServerConfig
 /// [`ClientConfig`]: crate::ClientConfig
-pub fn config_builder() -> ConfigWantsCipherSuites {
-    ConfigWantsCipherSuites {}
+/// [`ClientConfig::builder()`]: crate::ClientConfig::builder()
+/// [`ServerConfig::builder()`]: crate::ServerConfig::builder()
+/// [`ConfigWantsServerVerifier`]: crate::ConfigWantsServerVerifier
+/// [`ConfigWantsClientVerifier`]: crate::ConfigWantsClientVerifier
+#[derive(Clone)]
+pub struct ConfigBuilder<Side: ConfigSide, State> {
+    pub(crate) state: State,
+    pub(crate) side: PhantomData<Side>,
 }
 
-fn config_builder_with_safe_defaults() -> ConfigWantsPeerType {
-    config_builder()
-        .with_safe_default_cipher_suites()
-        .with_safe_default_kx_groups()
-        .with_safe_default_protocol_versions()
-}
+/// Config builder state where the caller must supply cipher suites.
+#[derive(Clone)]
+pub struct WantsCipherSuites;
```

Comment:
nit, take it or leave it: this will allow users to publicly construct this type, and means adding any fields in this state is a breaking change. although it won't really be useful to publicly construct this, i think it might be worth adding a private field so user code can't construct it, especially in case it becomes necessary to add fields later?
```suggestion
pub struct WantsCipherSuites {
   _private: (),
}
```